README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- badges: start -->
<!-- badges: end -->
The \`\`genodds'' package <img src="sticker.png" alt="genodds logo" style="float:right;height:232.25px" align="right" height="232.25">
======================================================================================================================================

Overview
--------

genodds implements Agresti's [generalized odds ratios](https://www.jstor.org/stable/2530495) for a two-sample dataset. This measure calculates the odds that, if a pair of observation were to be randomly selected from two groups, the observation in one group would have a higher score than those in the other group. The package also provides hypothesis testing for if this odds ratio is different to 1 (no difference between groups), as well as power/sample size estimation capabilities trial planning using this test. This measure can also be reported as [Number Needed to Treat](https://doi.org/10.1111/ijs.12364), a common outcome measure used in health economics.

Installation
------------

You can install the released version of genodds from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("genodds")
```

Example
-------

genodds can be run as follows:

``` r
library(genodds)
df <- alteplase
genodds(df$mRS,df$treat,df$time)
#>   Agresti's Generalised odds ratios
#> 
#>   0-90         Odds: 0.772 (0.599, 0.996)      p=0.0468
#>   91-180       Odds: 0.850 (0.709, 1.020)      p=0.0805
#>   181-270      Odds: 0.862 (0.771, 0.963)      p=0.0088
#>   271-360      Odds: 1.019 (0.891, 1.164)      p=0.7868
#> 
#> Test of H0: odds ratios are equal among strata:
#>   X-squared = 5.63, df= 3     p=0.1311
#> 
#> Test of H0: pooled odds = 1:
#>   Pooled odds: 0.897 (0.833,0.966)  p=0.0039
```


NEWS.md
# genodds 1.1.2

Added useful contributions by Howard on a related approach. Method now also offers a permutation test equivalent method, plus an interpretable statement of effect as proposed by Howard.

Patched a bug when switching between dropping and splitting ties. Previously, when assuming null standard error and dropping ties confidence intervals and p-values would erroneously return NA values.

# genodds 1.1

Added option for controlling standard error. The use of standard error under Null for calculations (Following O'Brien's approach is now achieved by means of the assume_no_effect option).

Patched a bug in the calculation of cell proportions for the purposes of calculating standard error under null. Previous version assumed equal group sizes.

NOTE: I forgot to upload this to Github for the dev version. See cran/genodds for this version.

# genodds 1.0

Package released


tests/testthat.R
library(testthat)
library(genodds)

test_check("genodds")


tests/testthat/test-genodds.R
context("Input Sanitation for genodds()")

test_that("Error if response is empty", {
  expect_error(
    genodds(NULL,alteplase$treat)
  )
})

test_that("Error if group is empty", {
  expect_error(
    genodds(alteplase$mRS,NULL)
  )
})


test_that("Error if response and group are of different length", {
  expect_error(
    genodds(alteplase$mRS[1:100],alteplase$treat)
  )
})


test_that("Error if response and strata are of different length", {
  expect_error(
    genodds(alteplase$mRS,alteplase$treat,alteplase$time[100])
  )
})

test_that("Error if treat contains more than two groups", {
  df <- alteplase
  df$treat <- as.character(df$treat)
  df$treat[c(100,600,651)] <- "Other"

  expect_error(
    genodds(df$mRS,df$treat,df$time)
  )
})

test_that("Warning if response contains NA", {
  df <- alteplase
  df$mRS[c(100,600,651)] <- NA

  expect_warning(
    genodds(df$mRS,df$treat,df$time)
  )
})

test_that("Warning if group contains NA", {
  df <- alteplase
  df$treat[c(100,600,651)] <- NA

  expect_warning(
    genodds(df$mRS,df$treat,df$time)
  )
})


test_that("Warning if strata contains NA", {
  df <- alteplase
  df$time[c(100,600,651)] <- NA

  expect_warning(
    genodds(df$mRS,df$treat,df$time)
  )
})


context("Calculation tests for genodds()")

test_that("pooled odds equal to stratum odds when stratum=NULL", {

  obj <- genodds(alteplase$mRS,alteplase$treat)

  expect_equal(exp(obj$pooled_lnodds), obj$results[[1]]$odds)
})

test_that("pooled p value equal to stratum p when stratum=NULL", {

  obj <- genodds(alteplase$mRS,alteplase$treat)

  expect_equal(obj$pooled_p, obj$results[[1]]$p)
})


context("Input Sanitation for genodds.power()")

test_that("Error if neither N nor power are specified",{

          p0 <- c(0.224,0.191,0.082,0.133,0.136,0.043,0.191)
          p1 <- c(0.109,0.199,0.109,0.120,0.194,0.070,0.200)

          expect_error(
            genodds.power(p0,p1)
          )
          })


test_that("Error if power isn't in [0,1]",{

  p0 <- c(0.224,0.191,0.082,0.133,0.136,0.043,0.191)
  p1 <- c(0.109,0.199,0.109,0.120,0.194,0.070,0.200)

  expect_error(
    genodds.power(p0,p1,power=-1)
  )

  expect_error(
    genodds.power(p0,p1,power=2)
  )

})

test_that("Error if N isn't >=2",{

  p0 <- c(0.224,0.191,0.082,0.133,0.136,0.043,0.191)
  p1 <- c(0.109,0.199,0.109,0.120,0.194,0.070,0.200)

  expect_error(
    genodds.power(p0,p1,N=1)
  )
})

test_that("Error p0 and p1 are different lengths",{

  p0 <- c(0.224,0.191,0.082,0.133,0.136,0.043,0.191)
  p1 <- c(0.109,0.199,0.109,0.120,0.194,0.070)

  expect_error(
    genodds.power(p0,p1,N=1)
  )
})


context("Calculation Tests for genodds.power()")

test_that("Required sample size is conservative", {

  p0 <- c(0.224,0.191,0.082,0.133,0.136,0.043,0.191)
  p1 <- c(0.109,0.199,0.109,0.120,0.194,0.070,0.200)
  power_in <- seq(0.1,0.9,0.1)
  N <- genodds.power(p0,p1,power=power_in)

  power_out <- genodds.power(p0,p1,N=N)

  expect_true(prod(power_out>power_in)==1)
})

test_that("Power at sample size within 5e-2 of input power", {

  p0 <- c(0.224,0.191,0.082,0.133,0.136,0.043,0.191)
  p1 <- c(0.109,0.199,0.109,0.120,0.194,0.070,0.200)
  power_in <- seq(0.1,0.9,0.1)
  N <- genodds.power(p0,p1,power=power_in)

  power_out <- genodds.power(p0,p1,N=N)

  expect_true(prod((power_out-power_in)<5e-2)==1)
})

test_that("Theoretical power within 1% of empirical power", {

  set.seed(1908141806)

  p0 <- c(0.224,0.191,0.082,0.133,0.136,0.043,0.191)
  p1 <- c(0.109,0.199,0.109,0.120,0.194,0.070,0.200)

  power_in <- 0.8
  alpha=0.05
  N <- genodds.power(p0,p1,power=power_in)

  empiricalPower <- sapply(1:1e4,function(x){
    response <- c(
                  sample(0:6,size = ceiling(N/2),prob = p0,replace=TRUE),
                  sample(0:6,size = ceiling(N/2),prob = p1,replace=TRUE)
    )

    group <- c(
      rep(0,ceiling(N/2)),
      rep(1,ceiling(N/2))
    )

    obj <- genodds(response,group,alpha=alpha)$pooled_p<=alpha
  })

  empiricalPower <- mean(empiricalPower)

  expect_lte(abs(power_in - empiricalPower),1e-2)
})



R/data.R
#' Effect of time to treatment on efficacy of rt-PA in treating acute stroke
#'
#' A dataset reconstructing reported outcomes after stroke
#' as reported in Lees et. al. (2010) which pooled the results
#' of several studies.
#'
#' @format A data frame with 3669 rows and 3 variables:
#' \describe{
#'   \item{time}{Time interval (minutes) between stroke onset and treatment}
#'   \item{treat}{Type of treatment received}
#'   \item{mRS}{Outcome at 3 months measured using the modified Rankin Scale}
#' }
#'
#' @references Lees, K. R., Bluhmki, E., Von Kummer, R., Brott, T. G.,
#' Toni, D., Grotta, J. C.,  Albers, G. W., Kaste, M., Marler, J. R.,
#' Hamilton, S. A., Tilley, B. C., Davis, S. M., Donnan, G. A., Hacke, W.
#'  (2010).
#' Time to treatment with intravenous alteplase and outcome in stroke:
#' an updated pooled analysis of ECASS, ATLANTIS, NINDS, and EPITHET trials.
#' \emph{The Lancet}, 375(9727), 1695-1703.
#'
"alteplase"


R/genodds.power.R
#' Power Calculations for Generalized Odds Ratios
#' @description Provides power analysis for Agresti's Generalized Odds Ratios.
#'
#' @param p0 A numeric vector contianing the probabilities in control group.
#' @param p1 A numeric vector contianing the probabilities in treatment group.
#' @param N  A numeirc vector containing total sample sizes.
#' @param power A numeric vector containing required  total sample size.
#' @param alpha Type 1 error.
#' @param ties A string specifying how ties should be treated.
#'             Should be equal to "split" 0.5 for WMW Odds,
#'             or "drop" for Agresti's GenOR.
#' @param w A numeric vector of length 2 specifying the relative weighting
#'          of sample size between treatment groups.
#' @param direction Direction for hypothesis test.
#'                  Must be one of \code{"two.sided"},\code{"upper.tail"} or \code{"lower.tail"}.
#'
#' @return
#' \describe{
#'     \item{If \code{power} is supplied}{A numeric vector containing required sample sizes to achieve specified powers.}
#'     \item{If \code{N} is supplied}{A numeric vector containing power at specified sample sizes.}
#' }
#'
#' @details
#'
#' See \code{\link{genodds}} for explanation of generalized odds ratios.
#'
#' \code{N} provides the total sample size.
#' Sample size per group can be calculated by \code{N*w/sum(w)}.
#'
#' When \code{power} is supplied, if no sufficient sample size is
#' found then this function will return \code{Inf}.
#'
#' @examples
#'
#' # Provide theoretical distributions of outcomes for each group
#' # Distributions taken from Lees et. al. (2010). See ?alteplase for a citation.
#' p0 <- c(0.224,0.191,0.082,0.133,0.136,0.043,0.191)
#' p1 <- c(0.109,0.199,0.109,0.120,0.194,0.070,0.200)
#'
#' # Calculate sample size required to achieve 80% and 90%
#' # power for these distributions
#' genodds.power(p0,p1,power=c(0.8,0.9))
#'
#' # genodds.power suggests a total sample size of 619 for 80% power.
#' # Round up to 620 for even sample size per group
#'
#' # Confirm these sample sizes lead to 80% and 90% power
#' genodds.power(p0,p1,N=c(620,830))
#'
#' @references
#' O'Brien, R. G., & Castelloe, J. (2006, March).
#' Exploiting the link between the Wilcoxon-Mann-Whitney test and a simple odds statistic.
#' In \emph{Thirty-first Annual SAS Users Group International Conference}.
#'
#' @export
genodds.power <- function(p0,p1,N=NULL,power=NULL,
                          alpha=0.05,ties="split",
                          w=c(0.5,0.5),
                          direction="two.sided")
{

  if(is.null(p0) | is.null(p1))
  {
    stop("p0 and p1 must not be empty")
  }


  if(length(p0) != length(p1))
  {
    stop("p0 and p1 must be the same length")
  }

  if(is.null(N) & !is.null(power))
  {
    if(prod((power <=1) & (power >=0))==0)
    {
      stop("Power must be in the interval [0,1]")
    }

    if(length(power)>1)
    {
      # If multiple power values were requested, calculate power
      # for each

      out <- sapply(power,function(p){
        genodds.power(p0=p0,
                      p1=p1,
                      N=NULL,
                      power=p,
                      alpha=alpha,
                      ties=ties,
                      w=w,
                      direction=direction)
        })

      return(out)
    }
    else
    {
      # Recursively call this function and use
      # Bisection method to find required sample size to achieve
      # specified power

      # Find initial boundaries for search by doubling the initial window
      # until the power at upper boundary is above specified power
      nLow <- 2
      nHigh <- 4
      while(genodds.power(p0,p1,nHigh,NULL,alpha,ties,w,direction)<power)
      {
        nLow <- nHigh
        nHigh <- nHigh*2

        if(is.infinite(nHigh))
        {
          # If we never find an appropriate boundary, return Inf
          return(Inf)
        }
      }

      root <- uniroot(function(x){
        genodds.power(p0,p1,x,NULL,alpha,ties,w,direction)-power
      },
      c(nLow,nHigh)
      )

      # Round required sample size up to nearest integer
      return(ceiling(root$root))
    }
  }
  else if(!is.null(N) & is.null(power))
  {
    if(prod(N>=2)==0)
    {
      stop("N must be at least 2")
    }

    if(ties=="split")
    {
      contr_fav=0.5
    }
    else if (ties=="drop")
    {
      contr_fav=NA
    }
    else
    {
      stop("Invalid ties option specified")
    }

    # Plug values into formula to get power

    # Get ~pij = P(Y_i=j | group==i)
    p <- cbind(p0/sum(p0),p1/sum(p1))

    # This is p, the joing probability P(group==i,Y_i==j)
    wp <- p*cbind(rep(w[1],nrow(p)),rep(w[2],nrow(p)))

    Rt=wp[,2:1]
    Rs=get_Rs(wp)
    Rd=get_Rd(wp)

    # Redistribute ties
    if(!is.na(contr_fav))
    {
      Rs=Rs+(1-contr_fav)*Rt
      Rd=Rd+contr_fav*Rt
    }

    Pc=sum(wp*Rs)
    Pd=sum(wp*Rd)


    odds=Pc/Pd

    SEodds=2/Pd*(sum(wp*(odds*Rd-Rs)^2)/N)^0.5
    SElnodds=SEodds/odds


    # Primary effect size measure
    d <- log(odds)/(sqrt(N)*SElnodds)



    # Smooth p across groups and do this again.
    # This code is WET as hell, but whatever

    p=apply(p,1,mean)
    p=cbind(p,p)

    # standardise w to sum to 1
    w <- w/sum(w)

    wp <- p*cbind(rep(w[1],nrow(p)),rep(w[2],nrow(p)))

    Rt=wp[,2:1]
    Rs=get_Rs(wp)
    Rd=get_Rd(wp)

    # Redistribute ties
    if(!is.na(contr_fav))
    {
      Rs=Rs+(1-contr_fav)*Rt
      Rd=Rd+contr_fav*Rt
    }

    Pc=sum(wp*Rs)
    Pd=sum(wp*Rd)

    SEnull=2/Pd*(sum(wp*(1*Rd-Rs)^2)/N)^0.5
    SElnnull=SEnull/1

    if(direction=="two.sided")
    {
      nc <-N*d^2
      out <- pchisq((SElnnull/SElnodds)^2 * qchisq(1-alpha,1,0),
                    1,nc,
                    lower.tail = FALSE)
    }
    else if(direction=="lower.tail")
    {
      out <- pnorm(SElnnull/SElnodds * qnorm(alpha) - d*sqrt(N),lower.tail =T)
    }
    else if(direction=="upper.tail")
    {
      out <- pnorm(SElnnull/SElnodds * qnorm(1-alpha) - d*sqrt(N),lower.tail =F)
    }
    else
    {
      stop("incorrect direction")
    }

    return(out)
  }
  else
  {
    stop("must provide exactly one of N or power")
  }
}


